src/win32/path_w32.c

```diff
@@ -303,3 +308,223 @@ char *git_win32_path_8dot3_name(const char *path)
 
 	return shortname;
 }
+
+#if !defined(__MINGW32__)
+int git_win32_path_dirload_with_stat(
+	const char *path,
+	size_t prefix_len,
+	unsigned int flags,
+	const char *start_stat,
+	const char *end_stat,
+	git_vector *contents)
+{
+	int error = 0;
+	git_path_with_stat *ps;
+	git_win32_path pathw;
+	DIR *dir;
+	int(*strncomp)(const char *a, const char *b, size_t sz);
+	size_t cmp_len;
+	size_t start_len = start_stat ? strlen(start_stat) : 0;
+	size_t end_len = end_stat ? strlen(end_stat) : 0;
+	size_t path_size = strlen(path);
+	const char *repo_path = path + prefix_len;
+	size_t repo_path_len = strlen(repo_path);
+	char work_path[PATH__MAX_UNC_LEN];
+	git_win32_path target;
+	size_t path_len;
+	int fMode;
+
+	if (!git_win32__findfirstfile_filter(pathw, path)) {
+		error = -1;
+		giterr_set(GITERR_OS, "Could not parse the path '%s'", path);
+		goto clean_up_and_exit;
+	}
+
+	strncomp = (flags & GIT_PATH_DIR_IGNORE_CASE) != 0 
+		       ? git__strncasecmp 
+		       : git__strncmp;
+
+	/* use of FIND_FIRST_EX_LARGE_FETCH flag in the FindFirstFileExW call could benefit perormance
+	 * here when querying large repositories on Windows 7 (0x0600) or newer versions of Windows.
+	 * doing so could introduce compatibility issues on older versions of Windows. */
+	dir = git__calloc(1, sizeof(DIR));
+	dir->h = FindFirstFileExW(pathw, FindExInfoBasic, &dir->f, FindExSearchNameMatch, NULL, 0);
+	dir->first = 1;
+	if (dir->h == INVALID_HANDLE_VALUE) {
+		error = -1;
+		giterr_set(GITERR_OS, "Could not open directory '%s'", path);
+		goto clean_up_and_exit;
+	}
+	
+	if (repo_path_len > PATH__MAX_UNC_LEN) {
+		error = -1;
+		giterr_set(GITERR_OS, "Could not open directory '%s'", path);
+		goto clean_up_and_exit;
+	}
+
+	memcpy(work_path, repo_path, repo_path_len);
+
+	while (dir) {
+		if (!git_path_is_dot_or_dotdotW(dir->f.cFileName)) {
+			path_len = git__utf16_to_8(work_path + repo_path_len, ARRAYSIZE(work_path) - repo_path_len, dir->f.cFileName);
+
+			work_path[path_len + repo_path_len] = '\0';
+			path_len = path_len + repo_path_len;
+
+			cmp_len = min(start_len, path_len);
+			if (!(cmp_len && strncomp(work_path, start_stat, cmp_len) < 0)) {
+				cmp_len = min(end_len, path_len);
+				if (!(cmp_len && strncomp(work_path, end_stat, cmp_len) > 0)) {
+					fMode = S_IREAD;
+
+					if (dir->f.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+						fMode |= S_IFDIR;
+					else
+						fMode |= S_IFREG;
+
+					if (!(dir->f.dwFileAttributes & FILE_ATTRIBUTE_READONLY))
+						fMode |= S_IWRITE;
+
+					ps = git__calloc(1, sizeof(git_path_with_stat) + path_len + 2);
```

Comment:
I just realized that this was a copy and paste of `lstat_w`...  I'm going to pull this out into a common function, because this is ugly enough once that I certainly don't want it all a second time.